src/Datadog.Trace/AsyncLocalScopeManager.cs

```diff
@@ -8,26 +9,47 @@ internal class AsyncLocalScopeManager : IScopeManager
 
         private readonly AsyncLocalCompat<Scope> _activeScope = new AsyncLocalCompat<Scope>();
 
+        private readonly List<IScopeListener> _scopeListeners = new List<IScopeListener>(); // TODO: This needs to be thread safe...
+
         public Scope Active => _activeScope.Get();
 
         public Scope Activate(Span span, bool finishOnClose)
         {
             var activeScope = _activeScope.Get();
             var scope = new Scope(activeScope, span, this, finishOnClose);
-            _activeScope.Set(scope);
+
             return scope;
         }
 
         public void Close(Scope scope)
         {
-            var current = _activeScope.Get();
+            foreach (IScopeListener listener in _scopeListeners)
+            {
+                listener.AfterScopeClosed();
+            }
 
+            var current = _activeScope.Get();
             if (current != null && current == scope)
             {
                 // if the scope that was just closed was the active scope,
                 // set its parent as the new active scope
                 _activeScope.Set(current.Parent);
             }
         }
+
+        public void AddScopeListener(IScopeListener listener)
+        {
+            _scopeListeners.Add(listener);
+        }
+
+        private void SetScope(Scope scope)
+        {
+            _activeScope.Set(scope);
+
+            foreach (IScopeListener listener in _scopeListeners)
+            {
+                listener.AfterScopeActivated();
```

Comment:
Same idea here with `listener.OnScopeStarted(Scope scope)`
We may want to wrap this as well.

Reply:
I renamed to `OnScopeActivated` to keep with the `active` state of the scope. Is that fine?